source/common/config/unified_mux/delta_subscription_state.cc

```diff
@@ -0,0 +1,233 @@
+            message.resources().size(), message.removed_resources().size());
+}
+
+void DeltaSubscriptionState::handleBadResponse(const EnvoyException& e, UpdateAck& ack) {
+  // Note that error_detail being set is what indicates that a DeltaDiscoveryRequest is a NACK.
+  ack.error_detail_.set_code(Grpc::Status::WellKnownGrpcStatus::Internal);
+  ack.error_detail_.set_message(Config::Utility::truncateGrpcStatusMessage(e.what()));
+  ENVOY_LOG(warn, "delta config for {} rejected: {}", type_url(), e.what());
+  callbacks().onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason::UpdateRejected, &e);
+}
+
+void DeltaSubscriptionState::handleEstablishmentFailure() {
+  callbacks().onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason::ConnectionFailure,
+                                   nullptr);
+}
+
+envoy::service::discovery::v3::DeltaDiscoveryRequest*
+DeltaSubscriptionState::getNextRequestInternal() {
+  auto* request = new envoy::service::discovery::v3::DeltaDiscoveryRequest;
+  request->set_type_url(type_url());
+  if (!any_request_sent_yet_in_current_stream_) {
+    any_request_sent_yet_in_current_stream_ = true;
+    // initial_resource_versions "must be populated for first request in a stream".
+    // Also, since this might be a new server, we must explicitly state *all* of our subscription
+    // interest.
+    for (auto const& [resource_name, resource_state] : resource_state_) {
+      // Populate initial_resource_versions with the resource versions we currently have.
+      // Resources we are interested in, but are still waiting to get any version of from the
+      // server, do not belong in initial_resource_versions. (But do belong in new subscriptions!)
+      if (!resource_state.isWaitingForServer()) {
+        (*request->mutable_initial_resource_versions())[resource_name] = resource_state.version();
+      }
+      // As mentioned above, fill resource_names_subscribe with everything, including names we
+      // have yet to receive any resource for unless this is a wildcard subscription, for which
+      // the first request on a stream must be without any resource names.
+      if (!wildcard_) {
+        names_added_.insert(resource_name);
+      }
+    }
+    // Wildcard subscription initial requests must have no resource_names_subscribe.
+    if (wildcard_) {
+      names_added_.clear();
+    }
+    names_removed_.clear();
+  }
+
+  std::copy(names_added_.begin(), names_added_.end(),
+            Protobuf::RepeatedFieldBackInserter(request->mutable_resource_names_subscribe()));
+  std::copy(names_removed_.begin(), names_removed_.end(),
+            Protobuf::RepeatedFieldBackInserter(request->mutable_resource_names_unsubscribe()));
+  names_added_.clear();
+  names_removed_.clear();
+
+  return request;
+}
+
+envoy::service::discovery::v3::DeltaDiscoveryRequest*
+DeltaSubscriptionState::getNextRequestAckless() {
+  return getNextRequestInternal();
```

Comment:
Should be refactored and implemented at the parent as the SotW variant should do the same thing.

Reply:
done